qutip/solver/flimesolve.py

```diff
@@ -59,107 +59,103 @@ def _floquet_rate_matrix(floquet_basis,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-
-            flime_SecondTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ac,md,pb,qn->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-
-            flime_ThirdTerm = c_op_rates[cdx] \
-                * np.einsum('abcdlk,ac,nb,pm,qd->mnpq',
-                            valid_c_op_products,
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim),
-                            np.eye(Hdim, Hdim))
-            try:
-                total_R_tensor[key] += np.reshape(flime_FirstTerm - (1 / 2)
-                                                  * (flime_SecondTerm +
-                                                     flime_ThirdTerm),
-                                                  (Hdim**2, Hdim**2))
-            except KeyError:
-                total_R_tensor[key] = np.reshape(flime_FirstTerm - (1 / 2)
-                                                 * (flime_SecondTerm +
-                                                    flime_ThirdTerm),
-                                                 (Hdim**2, Hdim**2))
+
+        for l, k in product(np.arange(Nt), repeat=2):
+            delta_shift = delta_m + (l - k)
+            mask = {}
+            if time_sense <= 0.:
+                mask[0] = delta_shift == 0
+                if not np.any(mask):
+                    continue
+                rate_products = np.multiply.outer(
+                    c_op_Fourier_amplitudes_list[l],
+                    np.conj(c_op_Fourier_amplitudes_list)[k]
+                ) * c_op_rates[cdx]
+            else:
+                rate_products = np.multiply.outer(
+                    c_op_Fourier_amplitudes_list[l],
+                    np.conj(c_op_Fourier_amplitudes_list)[k]
+                )
+                cmp_array = np.minimum(np.abs(rate_products) * time_sense, Nt/2)
+
+
+                included_deltas = np.abs(delta_shift) * omega <= cmp_array
+                if not np.any(included_deltas):
+                    continue
+                keys = np.unique(delta_shift[included_deltas])
+                rate_products *= c_op_rates[cdx]
+                for key in keys:
+                    mask[key] = np.logical_and(delta_shift == key, included_deltas)
+
+            for key in mask.keys():
+                valid_c_op_products = rate_products * mask[key]
+                I_ = np.eye(Hdim, Hdim)
+
+                # using c = ap,d = bp,k=lp
+                flime_FirstTerm = np.transpose(valid_c_op_products, [0,2,1,3]).reshape(Hdim**2, Hdim**2)
```

Comment:
`np.einsum` does not know that the `eye` matrix is diagonal, it always do the full tensor product. Anytime the input is `eye`, there are probably a specialized function for it which reduce the number of operation from `~N**2` to `~N`.